From ef40dab47ba94ca47e6a268853d415d9311aca04 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Wed, 9 Jul 2025 09:38:50 +0100
Subject: [PATCH] vddk: Remove -io vddk-noextents option

Commit 191b8cf418 ("input: Add undocumented -io vddk-noextents=true
option") added an option which could be used to insert the nbdkit
noextents filter on top of the VDDK source.  This was designed so we
could test in production if this improved performance (which it
didn't).

Since then we've spent more time investigating problems with
QueryAllocatedBlocks and believe we have a better solution in nbdkit.
See:
https://gitlab.com/nbdkit/nbdkit/-/commit/5a882e74cae3dbaa09bf3b942a02f9947b12f6e5

Reverts: commit 191b8cf418076ae3766b134ffa96eee048c7eb9d
---
 README              |  1 -
 input/input_vddk.ml | 35 -----------------------------------
 2 files changed, 36 deletions(-)

diff --git a/README b/README
index 49082a96..1cdbef15 100644
--- a/README
+++ b/README
@@ -71,7 +71,6 @@ REQUIREMENTS
   + nbdkit-blocksize-filter
   + nbdkit-cow-filter
   + nbdkit-multi-conn-filter
-  + nbdkit-noextents-filter
   + nbdkit-rate-filter
   + nbdkit-retry-filter
 
diff --git a/input/input_vddk.ml b/input/input_vddk.ml
index 39356129..12f43729 100644
--- a/input/input_vddk.ml
+++ b/input/input_vddk.ml
@@ -58,7 +58,6 @@ All other settings are optional:
   -io vddk-file=FILE           Override nbdkit-vddk-plugin file= parameter
   -io vddk-libdir=LIBDIR       VDDK library parent directory
   -io vddk-nfchostport=PORT    VDDK nfchostport
-  -io vddk-noextents=true      Avoid slow VDDK QueryAllocatedBlocks API
   -io vddk-port=PORT           VDDK port
   -io vddk-snapshot=SNAPSHOT-MOREF
                                VDDK snapshot moref
@@ -79,7 +78,6 @@ information on these settings.
         "file";
         "libdir";
         "nfchostport";
-        "noextents";
         "port";
         "snapshot";
         "thumbprint";
@@ -208,9 +206,6 @@ information on these settings.
       try Some (List.assoc "libdir" io_options) with Not_found -> None in
     let nfchostport =
       try Some (List.assoc "nfchostport" io_options) with Not_found -> None in
-    let noextents =
-      try bool_of_string (List.assoc "noextents" io_options)
-      with Not_found -> false in
     let port =
       try Some (List.assoc "port" io_options) with Not_found -> None in
     let snapshot =
@@ -318,36 +313,6 @@ See also the virt-v2v-input-vmware(1) manual.") libNN
        *)
       Nbdkit.add_filter_if_available cmd "retry";
 
-      (* VDDK's QueryAllocatedBlocks API is infamously slow.  It appears
-       * to block all other requests while it is running.  This API is
-       * also only called during the copy phase, not during conversion
-       * (or if it is, extremely rarely).
-       *
-       * If fstrim was successful, then trimmed blocks are stored in
-       * the COW filter (see below), and so requests for extents stop
-       * at that layer.  However for areas of the disk that fstrim
-       * thinks contain data, we still have to go through to VDDK to
-       * fetch extents.
-       *
-       * We could therefore add nbdkit-noextents-filter here (below COW,
-       * above VDDK plugin) which stops extents requests from going
-       * to VDDK, which would stop QueryAllocatedBlocks ever being
-       * called.  In my testing this is a moderate performance win.
-       *
-       * However ... in the case where fstrim failed, or for filesystems
-       * or partitions on the disk that we don't understand, doing this
-       * would mean that those are copied completely, as there would be
-       * no extent data (nbdcopy will still sparsify them on the target,
-       * but we'd have to copy all the bits from VMware).  Because
-       * here we don't know if this is the case, be conservative and
-       * actually don't use this filter.
-       *
-       * If used, this filter should be close to the plugin and MUST
-       * be below the COW filter.
-       *)
-      if noextents then
-        Nbdkit.add_filter_if_available cmd "noextents";
-
       (* Split very large requests to avoid out of memory errors on the
        * server.  Since we're using this filter, also add minblock=512
        * although it will make no difference.
